From d4958e97cbf43c5d1e6c16d33c8212b1500c5bc1 Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Tue, 11 Aug 2026 09:01:51 +0000
Subject: [PATCH] Try catch around constrainNitsAndBacklightArrays, and falls
 back to dumb curve. It crashes on Xperia 1 IV.

[Port to A16] Function body identical upstream; only trailing context moved (loadThermalThrottlingConfig follows loadBrightnessMap in A16).

Change-Id: I909a2d6b17a97df4cbdcff5636f7f361399f5146
---
 .../server/display/DisplayDeviceConfig.java   | 61 +++++++++++--------
 1 file changed, 34 insertions(+), 27 deletions(-)

diff --git a/services/core/java/com/android/server/display/DisplayDeviceConfig.java b/services/core/java/com/android/server/display/DisplayDeviceConfig.java
index 454fb2f807c4..096b4777eb99 100644
--- a/services/core/java/com/android/server/display/DisplayDeviceConfig.java
+++ b/services/core/java/com/android/server/display/DisplayDeviceConfig.java
@@ -2093,37 +2093,44 @@ public class DisplayDeviceConfig {
             return;
         }
 
-        // Use the (preferred) display device config mapping
-        final List<Point> points = map.getPoint();
-        final int size = points.size();
-
-        float[] nits = new float[size];
-        float[] backlight = new float[size];
-
-        mInterpolationType = convertInterpolationType(map.getInterpolation());
-        int i = 0;
-        for (Point point : points) {
-            nits[i] = point.getNits().floatValue();
-            backlight[i] = point.getValue().floatValue();
-            if (i > 0) {
-                if (nits[i] < nits[i - 1]) {
-                    Slog.e(TAG, "screenBrightnessMap must be non-decreasing, ignoring rest "
-                            + " of configuration. Nits: " + nits[i] + " < " + nits[i - 1]);
-                    return;
-                }
+        try {
+            // Use the (preferred) display device config mapping
+            final List<Point> points = map.getPoint();
+            final int size = points.size();
+
+            float[] nits = new float[size];
+            float[] backlight = new float[size];
+
+            mInterpolationType = convertInterpolationType(map.getInterpolation());
+            int i = 0;
+            for (Point point : points) {
+                nits[i] = point.getNits().floatValue();
+                backlight[i] = point.getValue().floatValue();
+                if (i > 0) {
+                    if (nits[i] < nits[i - 1]) {
+                        Slog.e(TAG, "screenBrightnessMap must be non-decreasing, ignoring rest "
+                                + " of configuration. Nits: " + nits[i] + " < " + nits[i - 1]);
+                        return;
+                    }
 
-                if (backlight[i] < backlight[i - 1]) {
-                    Slog.e(TAG, "screenBrightnessMap must be non-decreasing, ignoring rest "
-                            + " of configuration. Value: " + backlight[i] + " < "
-                            + backlight[i - 1]);
-                    return;
+                    if (backlight[i] < backlight[i - 1]) {
+                        Slog.e(TAG, "screenBrightnessMap must be non-decreasing, ignoring rest "
+                                + " of configuration. Value: " + backlight[i] + " < "
+                                + backlight[i - 1]);
+                        return;
+                    }
                 }
+                ++i;
             }
-            ++i;
+            mRawNits = nits;
+            mRawBacklight = backlight;
+            constrainNitsAndBacklightArrays();
+
+        } catch(Throwable t) {
+            mRawNits = null;
+            mRawBacklight = null;
+            setSimpleMappingStrategyValues();
         }
-        mRawNits = nits;
-        mRawBacklight = backlight;
-        constrainNitsAndBacklightArrays();
     }
 
     private void loadThermalThrottlingConfig(DisplayConfiguration config) {
-- 
2.43.0

